osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/messaging/DeviceRequestMessageProcessor.java

```diff
@@ -80,75 +87,93 @@ public void processMessage(final ObjectMessage message) throws JMSException {
 
         MessageMetadata messageMetadata = null;
-        DlmsConnectionManager conn = null;
+        final DlmsConnectionManager conn = null;
         DlmsDevice device = null;
 
         try {
             messageMetadata = MessageMetadata.fromMessage(message);
 
-            /**
-             * The happy flow for addMeter requires that the dlmsDevice does not
-             * exist. Because the findDlmsDevice below throws a runtime
-             * exception, we skip this call in the addMeter flow. The
-             * AddMeterRequestMessageProcessor will throw the appropriate
-             * 'dlmsDevice already exists' error if the dlmsDevice does exists!
-             */
-            if (!MessageType.ADD_METER.name().equals(messageMetadata.getMessageType())) {
+            if (this.requiresExistingDevice()) {
                 device = this.domainHelperService.findDlmsDevice(messageMetadata);
             }
 
-            LOGGER.info("{} called for device: {} for organisation: {}", message.getJMSType(),
-                    messageMetadata.getDeviceIdentification(), messageMetadata.getOrganisationIdentification());
-
-            final Serializable response;
-            if (this.usesDeviceConnection()) {
-                conn = this.createConnectionForDevice(device, messageMetadata);
-                response = this.handleMessage(conn, device, message.getObject());
-            } else {
-                response = this.handleMessage(device, message.getObject());
-            }
+            LOGGER.info("{} called for device: {} for organisation: {}, correlationUID={}", message.getJMSType(),
+                    messageMetadata.getDeviceIdentification(), messageMetadata.getOrganisationIdentification(),
+                    messageMetadata.getCorrelationUid());
 
-            // Send response
-            this.sendResponseMessage(messageMetadata, ResponseMessageResultType.OK, null, this.responseMessageSender,
-                    response);
+            final RequestWithMetadata<S> request = this.createRequestWithMetadata(message, messageMetadata);
+            final Serializable response = this.getResponse(conn, device, request);
+            this.sendResponse(messageMetadata, response);
         } catch (final JMSException exception) {
             this.logJmsException(LOGGER, exception, messageMetadata);
         } catch (final Exception exception) {
-            // Return original request + exception
-            if (!(exception instanceof SilentException)) {
-                LOGGER.error("Unexpected exception during {}", this.messageType.name(), exception);
-            }
-
-            this.sendResponseMessage(messageMetadata, ResponseMessageResultType.NOT_OK, exception,
-                    this.responseMessageSender, message.getObject());
+            this.sendErrorResponse(messageMetadata, exception, message.getObject());
         } finally {
             this.doConnectionPostProcessing(device, conn);
```

Comment:
Creation of the connection was moved to the getResponse() method. Now it is moved back outside this method. THe created connection (or null in case !usesDeviceConnection) is passed to the getResponse() method and passed to the doConnectionPostProcessing() in the finally block

```
    try {
      ...
      if (this.usesDeviceConnection()) {
        connectionManager = this.createConnectionForDevice(device, messageMetadata);
      }
      ...
      final Serializable response =
          this.getResponse(connectionManager, device, message.getObject(), messageMetadata);
      ...
    } finally {
      this.doConnectionPostProcessing(device, connectionManager);
    }

---

  protected Serializable getResponse(final DlmsConnectionManager connectionManager, ...) {
    if (connectionManager != null) {
      return this.handleMessage(connectionManager, device, request, messageMetadata);
    } else {
      return this.handleMessage(device, request, messageMetadata);
    }
  }
```